### ISSUE TITLE
Rule for removing unneeded parentheses

### ISSUE THREAD

**Issue by mvorisek** [PHP-CS-Fixer/PHP-CS-Fixer#4997]
Feature request for "remove_unneeded_parentheses".

Very important rule to allow to remove unneeded parentheses when:
- code is generated as generated code like:
   ```
   throw (new Exception(('string')));
   ```
- overused manually like:
   ```
   return ($v > 0) ? ($a + 1) : ($b - 1);
   ```

**Comment by kubawerlos** [comment]
Let's have some features merged first (as we have [31](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pulls?q=is%3Aopen+is%3Apr+label%3Akind%2Ffeature) opened feature pull requests, when I write this message) and then add [`NoUselessParenthesisFixer`](https://github.com/kubawerlos/php-cs-fixer-custom-fixers#NoUselessParenthesisFixer) here.

**Comment by mvorisek** [comment]
Is it now the good time? :)

**Comment by SpacePossum** [comment]
Sure ;)

How about making a `no_unneeded_parentheses` fixer with options so that `no_unneeded_control_parentheses` can be deprecated and be a proxy till we remove it on 3.0 in favor of the more generic new `no_unneeded_parentheses`?

I would also like to see the most basic ones cleaned up, like;  `echo 2 + foo(( 1 )) + 1;`

**Comment by mvorisek** [comment]
> How about making a `no_unneeded_parentheses` fixer with options so that `no_unneeded_control_parentheses` can be deprecated and be a proxy till we remove it on 3.0 in favor of the more generic new `no_unneeded_parentheses`?

I agree, now we can merge as proxy and possibly replace https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/d658dc0201bdd457d32fc412d810885f282ac064/src/RuleSet.php#L109 with the new `no_unneeded_parentheses` with BC settings).

In 3.0, we can change the Doctrine ruleset with more stricter/breaking settings.

**Comment by gharlan** [comment]
another example that could be fixed: 

```diff
- !($foo instanceof Foo)
+ !$foo instanceof Foo
```

**Comment by mvorisek** [comment]
> 
> 
> another example that could be fixed:
> 
> ```diff
> - !($foo instanceof Foo)
> + !$foo instanceof Foo
> ```

I agree, please submit as a test case to https://github.com/kubawerlos/php-cs-fixer-custom-fixers#NoUselessParenthesisFixer

**Comment by mvorisek** [comment]
@SpacePossum in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/6409 you write:

> some of the cases raised here https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4997 , but not all

what are the missing cases?

currently I see PHP CS Fixer can fix:

```
if (!$key instanceof AbstractScope) {
```

but not

```
} elseif (!is_string($key) && !($key instanceof Expressionable)) {
```

Is this a bug of some specific behaviour of `no_unneeded_control_parentheses` rule? If yes, I would prefer a separate bug ticket and close this broad issue instead.

**Comment by SpacePossum** [comment]
>if (!($object instanceof static)) {

Can be fixed using `no_unneeded_control_parentheses` using the `negative_instanceof` configuration option (I think that is what you are looking for).

The other case, with the ternary, is indeed not fixed. It is not a bug it just hasn't been made (and given how PHP ternary work it pretty hard, also because its prio isn't the same over all PHP versions).
Also note that the official PHP doc warns us about not grouping:
https://www.php.net/manual/en/language.operators.precedence.php

````
Operator precedence and associativity only determine how expressions are grouped, they do not specify an order of evaluation.
PHP does not (in the general case) specify in which order an expression is evaluated and code that assumes a specific order of
evaluation should be avoided,
because the behavior can change between versions of PHP or depending on the surrounding code. 
```

Other cases that are not fixed are about complex binary statements as these are also hard to check within the tool.

**Comment by mvorisek** [comment]
> > if (!($object instanceof static)) {
> 
> Can be fixed using `no_unneeded_control_parentheses` using the `negative_instanceof` configuration option (I think that is what you are looking for).

it seems it is not - reported in https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/6586

> The other case, with the ternary, is indeed not fixed. It is not a bug it just hasn't been made (and given how PHP ternary work it pretty hard, also because its prio isn't the same over all PHP versions).

we should probably not fix nested ternary/null coalescing, but `instanceof` has much higher precendence than ternary

**Comment by SpacePossum** [comment]
>we should probably not fix nested ternary/null coalescing, but instanceof has much higher precendence than ternary

I tried to say, ternary support I didn't make
instanceof I added support for

these are not connected in any other from what I know

**Comment by mvorisek** [comment]
never stale until implemented

**Comment by Wirone** [comment]
@kubawerlos are you still [open for migrating your custom fixer to the core](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/4997#issuecomment-657270502)? I ask only for permission, not for doing this - it can be done by others too.

**Comment by mvorisek** [comment]
Another case to be fixed:
```diff
-$origValue = ($this->{$k} ?? null);
+$origValue = $this->{$k} ?? null;
```

**Comment by mvorisek** [comment]
not stale
